documentation/general/program-scale-testing-requirements/hero-implementation-details.md

```diff
@@ -0,0 +1,261 @@
+
+
+### CPS
+
+CPS (connection per second) this is a metric that shows the slow path performance, and we can get both TCP and UDP values.
+
+For TCP we use IxLoad since it has full TCP stack that is very configurable and can simulate a lot of different scenarios.
+
+While the hero test calls for 6 TCP packets SYN/SYNACK/ACK/FIN/FINACK/ACK, we make use of HTTP as necessarily that runs over TCP and on the wire, we will end up with 7 packets for every connection. 
+
+PPS used for CPS test can be seen the L23 stats in IxLoad.
+
+Keep an eye on TCP failures on client and server a retransmit is bad it symbolizes packet drop that was detected and TCP stack had to retransmit. a connection drop is super extra bad it means even after 3-5 retries packet did not made it. 
+
+We also look at number of concurrent connection while the test is running. traffic generator puts on the wire equally time spaced SYN packets to match the desired CPS but rest of communication happens as fast as possible. impacted by line rate and latency. in theory if line rate is high and latency low the whole exchange of 7 packets could finish before the next SYN is sent resulting in 0 concurrent connection. (flow table will be 1), while a slow travel time for packets will results in connections that have not been terminated yet as new connections get initiated and this will result in a certain number concurrent connection. Ideally we want to see the concurrent connection number as low as possible.
+
+test tries to cycle through all the millions of IPs, source port is chosen at random in a specified range and destination port is fixed to 80
+
+we can do variations like witch side initiates the fin and see if we observe and differences in performance
+
+for UDP CPS slow path test use random source/dest ports and send a fixed amount of packets not exceeding the flow table size.
+
+Note down the bandwidth utilized by the CPS test.
+
+### Flow table size
+
+Flow timer must be set to a very high value so flows do not expire during the test
+
+For TCP we set the desired number of concurrent connection and make sure we have a transaction rate that is a bit faster than the flow timer to make sure flows do not expire.
+
+for UDP we use random source/destination ports and we set rate to 100K PPS and for 32M flows it should work fine for 320 seconds.
+
+we look here that flow table can be filed to desired level.
+
+one item to note here is to characterize what happens when flow table is full. will it crash? will it drop anything after? will it all extra packets be processed as slow path?
+
+### Background UDP flows
+
+
+### Background UDP flows
+
+
+### Hero test
+
+Putting it all together and running CPS test with background traffic.
+
+Start first the background traffic and ensure the flow table is close to full but not full (need room for CPS), increase packet size to ensure bandwith is utilized at 100% - bandwitch needed by CPS test - a 5%-10% margin
```

Comment:
s/bandwitch/bandwidth/

LoL "Band Witch" sounds like a disagreeable member (or partner thereof) in a rock band